Update biobakery tools to v3

boulund opened the pull request:
The work by @AroArz is starting to get into a mergeable state. I created this PR just to keep track of progress until we reach then end!

AroArz commented:
On my to-do for this PR.

1. Since merging with develop I've been getting cryptic error messages when running. I'm adhering these issues to 
1.1. TMPDIR suddenly not working as expected.
1.2. Problems with the conda and singularity envs.
2. Fix circle I.

boulund commented:
I encountered issues running metaphlan (v3) in a fresh conda environment:
```
(biobakery) $ metaphlan --install --bowtie2db /ceph/db/metaphlan3/

Downloading http://cmprod1.cibio.unitn.it/biobakery3/metaphlan_databases/mpa_latest
Downloading file of size: 0.00 MB
0.01 MB 31507.69 %  34.42 MB/sec  0 min -0 sec
Downloading MetaPhlAn database
Please note due to the size this might take a few minutes

Downloading http://cmprod1.cibio.unitn.it/biobakery3/metaphlan_databases/mpa_v30_CHOCOPhlAn_201901.tar
Downloading file of size: 366.62 MB
366.62 MB 100.00 %  47.95 MB/sec  0 min -0 sec
Downloading http://cmprod1.cibio.unitn.it/biobakery3/metaphlan_databases/mpa_v30_CHOCOPhlAn_201901.md5
Downloading file of size: 0.00 MB
0.01 MB 12800.00 %  45.13 MB/sec  0 min -0 sec

Decompressing /ceph/db/metaphlan3/mpa_v30_CHOCOPhlAn_201901.fna.bz2 into /ceph/db/metaphlan3/mpa_v30_CHOCOPhlAn_201901.fna
Traceback (most recent call last):
  File "/ceph/home/boulund/.conda/envs/biobakery/bin/metaphlan", line 10, in <module>
    sys.exit(main())
  File "/ceph/home/boulund/.conda/envs/biobakery/lib/python3.7/site-packages/metaphlan/metaphlan.py", line 925, in main
    pars['index'] = check_and_install_database(pars['index'], pars['bowtie2db'], pars['bowtie2_build'], pars['nproc'],
 pars['force_download'])
  File "/ceph/home/boulund/.conda/envs/biobakery/lib/python3.7/site-packages/metaphlan/__init__.py", line 301, in chec
k_and_install_database
    download_unpack_tar(index, bowtie2_db, bowtie2_build, nproc, use_zenodo)
  File "/ceph/home/boulund/.conda/envs/biobakery/lib/python3.7/site-packages/metaphlan/__init__.py", line 180, in down
load_unpack_tar
    bt2_build_output = subp.check_output([bowtie2_build, '--usage'], stderr=subp.STDOUT)
  File "/ceph/home/boulund/.conda/envs/biobakery/lib/python3.7/subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "/ceph/home/boulund/.conda/envs/biobakery/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['bowtie2-build', '--usage']' returned non-zero exit status 127.
```

It appears to be caused by an issue with `bowtie2-build`:
```
(biobakery) [boulund@c1hitachi10 metaphlan3]$ bowtie2-build
/ceph/home/boulund/.conda/envs/biobakery/bin/bowtie2-build-s: error while loading shared libraries: libtbb.so.2: cannot open shared object file: No such file or directory
```

Found this: https://forum.biobakery.org/t/workflow-conda-install-bowtie-2-issue-bowtie2-align-s-error-while-loading-shared-libraries-libtbb-so-2/1831

```
conda install tbb=2020.2
```

This downgraded the `tbb` package:
```
## Package Plan ##

  environment location: /ceph/home/boulund/.conda/envs/biobakery

  added / updated specs:
    - tbb=2020.2


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    tbb-2020.2                 |       h4bd325d_4         1.5 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         1.5 MB

The following packages will be DOWNGRADED:

  tbb                                   2021.2.0-h4bd325d_0 --> 2020.2-h4bd325d_4
```

This made it work:

```
(biobakery) $ metaphlan --install --bowtie2db /ceph/db/metaphlan3/

Downloading MetaPhlAn database
Please note due to the size this might take a few minutes

File /ceph/db/metaphlan3/mpa_v30_CHOCOPhlAn_201901.tar already present!

File /ceph/db/metaphlan3/mpa_v30_CHOCOPhlAn_201901.md5 already present!

Building Bowtie2 indexes
Removing uncompress database /ceph/db/metaphlan3/mpa_v30_CHOCOPhlAn_201901.fna

Download complete
The database is installed
```

boulund commented:
I wonder if there is a better way to handle messages to the user on runtime. We've been using print statements in the snakefiles but it becomes quite noisy in the terminal output when you are running, since all print statements are executed everytime snakemake reads the snakefiles... :thinking: 
Besides that, it looks pretty good! I will have to try to find some time to test properly.

boulund commented:
I have tested a run with humann and it seems to work as intended now. I made sure it cleans up the temporary folder after it's done. There's still a potential issue that the folder will not be removed if the process fails for some reason, but I'm hoping it won't create too many troubles.

boulund commented:
And now I've tested a run with StrainPhlAn as well. One more test, then I'm happy.

boulund commented:
I think I'm happy now @AroArz . Does that mean we are ready to merge?? :scream_cat:

boulund commented:
Yes, I was thinking about that as well... One alternative that I was considering is to just push potential warning/info messages into a list, and then in the `onerror:` part print everything in that list. Then it would be nicely summarized for the user if the workflow failed for some reason. The strainphlan errors would be a prime candidate to be handled like that.

boulund commented:
@AroArz I implemented a new class `UserMessages` that is instantiated as `user_messages`  in the main Snakefile scope. Messages added to this object are then printed in `onerror` and `onsuccess` handlers. I also moved the diagnostic printout about which samples were identified to the `onstart` handler.
